### PR TITLE
 [FLINK-23080][datastream] Introduce SinkFunction#finish() method. 

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change

Add an extension point for flushing buffered data in sinks before the final checkpoint.

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

**Comment by dawidwys** [comment]
Could you take a look @gaoyunhaii ?

**Comment by gaoyunhaii** [comment]
No problem, I'll have a look~

**Review by gaoyunhaii — COMMENTED** [review]
Very thanks @dawidwys  for opening the PR! The PR looks very good to me and I only have one small comment.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java [hunk 1]
```diff
@@ -211,6 +211,16 @@ protected void recoverAndAbort(TXN transaction) {
...
 
+    /**
+     * This method is called at the end of data processing.
+     *
+     * <p>The method is expected to flush all remaining buffered data. Exceptions will cause the
+     * pipeline to be recognized as failed, because the last data items are not processed properly.
+     * You may use this method to flush remaining buffered elements in the state into the current
+     * transaction which will be committed in the last checkpoint.
+     */
+    protected void finishProcessing(TXN transaction) {}
```
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java:222]: Sure.

**Pull request closed without merging**